README.md
# Erin

Parse tcpdump files with SMPP protocol and write it to Clickhouse.


Changes
v0.4.11 2019-03-18
  - Update to Go 1.12.
  - Remove errors logging for checked empty files.

v0.4.10 2018-12-12
  - Fix infinite move tries if file moved already.

v0.4.9 2018-11-19
  - Processing bugfix: incorrect read from Redis unexistant value.

v0.4.8 2018-11-18
  - Redis Cluster for store temporary data: final stage. Read data from
    Redis. Remove disk storage support.
    Remove env variable ERIN_TEMP_STORE_PATH.

v0.4.7 2018-11-15
  - Better start/stop logic.
  - Redis Cluster for store temporary data: migration period, only store data.
  - New Redis configuration variables.
  - Switch to Ami/Ruthie (Redis Cluster Streams based queues).
    New configuration variables: ERIN_SHARDS_COUNT, ERIN_PENDING_BUFFER_SIZE,
    ERIN_PIPE_BUFFER_SIZE.
    Remove RabbitMQ configuration variables: ERIN_RABBITMQ_USER,
    ERIN_RABBITMQ_PASSWORD, ERIN_RABBITMQ_ADDR, ERIN_RABBITMQ_VHOST.

v0.4.6 2018-10-03
  - Close file handlers.

v0.4.5 2018-10-01
  - Old files with ext.

v0.4.4 2018-09-08
  - Update dependencies.
  - Fix datetime of records in ClickHouse.

v0.4.3 2018-09-02
  - Bugfix old file store.

v0.4.2 2018-09-02
  - Timezone set fix.

v0.4.1 2018-09-01
  - Docker run fix.

v0.4.0 2018-09-01
  - Split old files to folders.
  - Change dump filenames.

v0.3.0 2018-09-01
  - Create work dirs.
  - Renamed variables.

v0.2.0 2018-09-01
  - Fix table structure.

v0.1.0 2018-07-18
  - First release.


parser/old.go
package parser

import (
	"os"
	"path/filepath"
	"time"
)

func (p *Parser) moveOld() {
	now := time.Now()

	for file := range p.toMove {
		// Check if file moved already
		_, err := os.Stat(file)
		if err != nil && os.IsNotExist(err) {
			delete(p.toMove, file)
			continue
		}

		_, name := filepath.Split(file)

		// connectionName_connectionID_20060102_150405.pcap
		timeFromName := name[len(name)-20 : len(name)-5]

		tm, err := time.ParseInLocation("20060102_150405", timeFromName, time.Local)
		if err != nil {
			p.logger.Error(err)
			continue
		}

		if now.Sub(tm).Hours() <= 2 {
			continue
		}

		noTime := name[0 : len(name)-21]

		part1 := tm.Format("20060102")
		part2 := tm.Format("150405") + ".pcap"

		newDir := filepath.Join(p.config.MovePath, noTime, part1)

		err = os.MkdirAll(newDir, 0775)
		if err != nil {
			p.logger.Error(err)
			continue
		}

		newPath := filepath.Join(newDir, part2)

		err = os.Rename(file, newPath)
		if err != nil {
			p.logger.Error(err)
			continue
		}

		delete(p.toMove, file)
	}
}


parser/parser.go
package parser

import (
	"strings"
	"sync"
	"time"

	"git.aqq.me/go/app/appconf"
	"git.aqq.me/go/app/applog"
	"git.aqq.me/go/app/event"
	"git.aqq.me/go/retrier"
	"github.com/go-redis/redis"
	"github.com/kak-tus/ami"
)

var prs *Parser

func init() {
	event.Init.AddHandler(
		func() error {
			cnfMap := appconf.GetConfig()["parser"]

			var cnf parserConfig
			err := appconf.Decode(cnfMap, &cnf)
			if err != nil {
				return err
			}

			loc, err := time.LoadLocation("UTC")
			if err != nil {
				return err
			}

			addrs := strings.Split(cnf.Redis.Addrs, ",")

			ropt := &redis.ClusterOptions{
				Addrs:    addrs,
				Password: cnf.Redis.Password,
			}

			redisdb := redis.NewClusterClient(ropt)

			pr, err := ami.NewProducer(
				ami.ProducerOptions{
					Name:              cnf.QueueName,
					ShardsCount:       cnf.ShardsCount,
					PendingBufferSize: cnf.PendingBufferSize,
					PipeBufferSize:    cnf.PipeBufferSize,
					PipePeriod:        time.Microsecond * 10,
				},
				ropt,
			)
			if err != nil {
				return err
			}

			prs = &Parser{
				logger:   applog.GetLogger().Sugar(),
				m:        &sync.Mutex{},
				C:        make(chan string, 1000),
				toMove:   make(map[string]bool),
				config:   cnf,
				location: loc,
				redisdb:  redisdb,
				retrier:  retrier.New(retrier.Config{RetryPolicy: []time.Duration{time.Second * 5}}),
				pr:       pr,
			}

			prs.logger.Info("Started parser")

			return nil
		},
	)

	event.Stop.AddHandler(
		func() error {
			prs.logger.Info("Stop parser")
			prs.m.Lock()

			prs.pr.Close()

			err := prs.redisdb.Close()
			if err != nil {
				return err
			}

			prs.logger.Info("Stopped parser")
			return nil
		},
	)
}

// GetParser return instance
func GetParser() *Parser {
	return prs
}

// Start parser
func (p *Parser) Start() {
	p.m.Lock()

	for {
		file, more := <-p.C

		if !more {
			break
		}

		p.process(file)
	}

	p.m.Unlock()
}

// Stop parser
func (p *Parser) Stop() {
	close(p.C)
}


founder/founder.go
package founder

import (
	"os"
	"path/filepath"
	"regexp"
	"sync"

	"git.aqq.me/go/app/appconf"
	"git.aqq.me/go/app/applog"
	"git.aqq.me/go/app/event"
	"github.com/fsnotify/fsnotify"
	"github.com/kak-tus/erin/parser"
)

var fnd *Founder

func init() {
	event.Init.AddHandler(
		func() error {
			cnfMap := appconf.GetConfig()["founder"]

			var cnf founderConfig
			err := appconf.Decode(cnfMap, &cnf)
			if err != nil {
				return err
			}

			wtch, err := fsnotify.NewWatcher()
			if err != nil {
				return err
			}

			fnd = &Founder{
				logger:  applog.GetLogger().Sugar(),
				config:  cnf,
				watcher: wtch,
				parser:  parser.GetParser(),
				regexp:  regexp.MustCompile(cnf.Regexp),
				m:       &sync.Mutex{},
			}

			fnd.logger.Info("Started founder")

			return nil
		},
	)

	event.Stop.AddHandler(
		func() error {
			fnd.logger.Info("Stop founder")
			fnd.watcher.Close()
			fnd.m.Lock()

			fnd.parser.Stop()
			fnd.logger.Info("Stopped founder")
			return nil
		},
	)
}

// GetFounder return instance
func GetFounder() *Founder {
	return fnd
}

// Start find files
func (f *Founder) Start() {
	go f.parser.Start()

	f.m.Lock()

	f.findInitial()

	err := f.watcher.Add(f.config.DumpPath)
	if err != nil {
		f.logger.Panic(err)
	}

LOOP:
	for {
		select {
		case ev, more := <-f.watcher.Events:
			if !more {
				break LOOP
			}

			if !(ev.Op == fsnotify.Create || ev.Op == fsnotify.Write) {
				continue
			}

			st, err := os.Stat(ev.Name)
			if err != nil {
				f.logger.Error(err)
				continue
			}

			// Dump file header has 4 bytes
			if st.Size() < 4 {
				continue
			}

			matchPattern, err := filepath.Match(filepath.Join(f.config.DumpPath, f.config.Pattern), ev.Name)
			if err != nil {
				f.logger.Error(err)
				continue
			}

			if matchPattern && f.matchRegexp(ev.Name) {
				f.parser.C <- ev.Name
			}
		case err, more := <-f.watcher.Errors:
			if !more {
				break LOOP
			}

			f.logger.Error(err)
		}
	}

	f.m.Unlock()
}

func (f *Founder) findInitial() {
	list, err := filepath.Glob(filepath.Join(f.config.DumpPath, f.config.Pattern))
	if err != nil {
		f.logger.Error(err)
		return
	}

	for _, file := range list {
		if f.matchRegexp(file) {
			f.parser.C <- file
		}
	}
}

func (f *Founder) matchRegexp(file string) bool {
	// connectionName_connectionID_20060102_150405.pcap
	_, name := filepath.Split(file)

	if len(name) < 24 {
		return false
	}

	return f.regexp.MatchString(name)
}


parser/process.go
package parser

import (
	"bytes"
	"crypto/sha1"
	"fmt"
	"path/filepath"
	"strconv"
	"strings"

	"git.aqq.me/go/retrier"
	"github.com/fiorix/go-smpp/smpp/pdu"
	"github.com/fiorix/go-smpp/smpp/pdu/pdufield"
	"github.com/fiorix/go-smpp/smpp/pdu/pdutext"
	"github.com/go-redis/redis"
	"github.com/google/gopacket"
	"github.com/google/gopacket/pcap"
	"github.com/kak-tus/ruthie/message"
)

const sql = "INSERT INTO gate.smpp " +
	"(dt_part,dt,connection_id,operation,type,status,sequence_id,message_id,phone,short_number,txt_id) " +
	"VALUES (?,?,?,?,?,?,?,?,?,?,?);"

const sqlText = "INSERT INTO gate.smpp_texts (dt_part,id,txt) VALUES (?,?,?);"

func (p *Parser) process(file string) {
	p.logger.Debug("Got file ", file)
	p.toMove[file] = true

	p.parse(file)

	p.moveOld()
}

func (p *Parser) parse(file string) {
	_, name := filepath.Split(file)

	offset := 0

	rdKey := fmt.Sprintf("{erin_%s}", name)

	p.retrier.Do(func() *retrier.Error {
		val, err := p.redisdb.Get(rdKey).Int()

		if err == redis.Nil {
			return nil
		} else if err != nil {
			p.logger.Error(err)
			return retrier.NewError(err, false)
		}

		offset = val

		return nil
	})

	// connectionName_connectionID_20060102_150405.pcap
	noTime := name[0 : len(name)-21]
	str := noTime[strings.LastIndex(noTime, "_")+1:]
	connectionID, err := strconv.Atoi(str)
	if err != nil {
		p.logger.Error(err)
		return
	}

	hdl, err := pcap.OpenOffline(file)
	if err != nil {
		p.logger.Error(err)
		return
	}

	defer hdl.Close()

	src := gopacket.NewPacketSource(hdl, hdl.LinkType())
	count := 0

	for pac := range src.Packets() {
		count++

		if offset >= count {
			continue
		}

		appl := pac.ApplicationLayer()
		if appl == nil {
			continue
		}

		rdr := bytes.NewReader(appl.LayerContents())

		bdy, err := pdu.Decode(rdr)
		if err != nil {
			continue
		}

		data := make([]interface{}, 11)

		meta := pac.Metadata()
		header := bdy.Header()
		fields := bdy.Fields()

		inGMT := meta.Timestamp.In(p.location)

		data[0] = inGMT.Format("2006-01-02")          // dt_part
		data[1] = inGMT.Format("2006-01-02 15:04:05") // dt
		data[2] = connectionID                        // connection_id
		data[3] = header.ID.String()                  // operation
		data[4] = getMsgType(header, fields)          // type
		data[5] = header.Status                       // status
		data[6] = header.Seq                          // sequence_id

		// message_id
		messageID := fields["message_id"]
		if messageID != nil {
			data[7] = messageID.String()
		}

		var phone pdufield.Body
		var shortNumber pdufield.Body

		if header.ID == pdu.SubmitSMID {
			phone = fields["destination_addr"]
			shortNumber = fields["source_addr"]
		} else if header.ID == pdu.DeliverSMID {
			phone = fields["source_addr"]
			shortNumber = fields["destination_addr"]
		}

		// phone
		if phone != nil {
			data[8] = phone.String()
		}

		// short_number
		if shortNumber != nil {
			data[9] = shortNumber.String()
		}

		txt := getText(header, fields)

		// txt_id
		if txt != nil {
			sha := sha1.Sum([]byte(txt.(string)))
			data[10] = string(sha[:])

			dataText := make([]interface{}, 3)

			dataText[0] = data[0]  // dt_part
			dataText[1] = data[10] // id
			dataText[2] = txt      // txt

			body, err := message.Message{
				Query: sqlText,
				Data:  dataText,
			}.Encode()

			if err == nil {
				p.pr.Send(body)
			}
		}

		body, err := message.Message{
			Query: sql,
			Data:  data,
		}.Encode()
		if err != nil {
			continue
		}

		p.pr.Send(body)
	}

	p.retrier.Do(func() *retrier.Error {
		status := p.redisdb.Set(rdKey, count, redisTTL)
		if status.Err() != nil {
			p.logger.Error(status.Err())
			return retrier.NewError(status.Err(), false)
		}

		return nil
	})

	p.logger.Debug("Processing done ", file)
}

func getText(header *pdu.Header, fields pdufield.Map) interface{} {
	if !(header.ID == pdu.SubmitSMID || header.ID == pdu.DeliverSMID) {
		return nil
	}

	if header.ID == pdu.DeliverSMID {
		esmClass, ok := fields["esm_class"]

		if ok && esmClass.String() == "4" {
			return nil
		}
	}

	txt, ok := fields["short_message"]
	if !ok {
		return nil
	}

	if len(txt.Bytes()) == 0 {
		return nil
	}

	coding, ok := fields["data_coding"]
	if !ok {
		return string(pdutext.GSM7(txt.Bytes()).Decode())
	}

	raw, err := strconv.Atoi(coding.String())
	if err != nil {
		return nil
	}

	c := pdutext.DataCoding(raw)

	var enc []byte

	if c == pdutext.ISO88595Type {
		enc = pdutext.ISO88595(txt.Bytes()).Decode()
	} else if c == pdutext.Latin1Type {
		enc = pdutext.Latin1(txt.Bytes()).Decode()
	} else if c == pdutext.UCS2Type {
		enc = pdutext.UCS2(txt.Bytes()).Decode()
	} else {
		enc = pdutext.GSM7(txt.Bytes()).Decode()
	}

	return string(enc)
}

func getMsgType(header *pdu.Header, fields pdufield.Map) interface{} {
	if header.ID == pdu.SubmitSMID {
		return "sms"
	} else if header.ID == pdu.DeliverSMID {
		esmClass, ok := fields["esm_class"]

		if ok && esmClass.String() == "4" {
			return "dlr"
		}

		return "sms"
	}

	return nil
}
